WalletWasabi/Crypto/ZeroKnowledge/ProofSystem.cs

```diff
@@ -40,5 +41,35 @@ public static Statement IssuerParameters(CoordinatorParameters iparams, MAC mac,
 				{ Generators.GV - iparams.I, O,             O,              Generators.Gx0, Generators.Gx1, Generators.Ga },
 				{ iparams.Cw,                Generators.Gw, Generators.Gwp, O,              O,              O },
 			});
+
+		public static Knowledge SerialNumber(GroupElement Ca, GroupElement S, Scalar z, Scalar a, Scalar r)
+			=> new Knowledge(SerialNumber(Ca, S), new ScalarVector(z, a, r));
+
+		public static Statement SerialNumber(GroupElement Ca, GroupElement S)
+			=> new Statement(new GroupElement[,]
+			{
+				// public                    Witness terms:
+				// point      z              a               r
+				{ Ca,         Generators.Ga, Generators.Gg,  Generators.Gh },
+				{ S,          O,             O,              Generators.Gs }
+			});
+
+		public static (GroupElement Ca, GroupElement S) SerialNumberPublicPoints(Scalar z, Scalar a, Scalar r)
+			=> (z * Generators.Ga + r * Generators.Gh + a * Generators.Gg, r * Generators.Gs);
+
+		public static Knowledge MacShow(CoordinatorParameters iparams, RandomizedCommitments c, Scalar z, Scalar t)
+			=> new Knowledge(MacShow(iparams, z * iparams.I, c.Cx0, c.Cx1), new ScalarVector(z, (t * z).Negate(), t));
+
+		public static Statement MacShow(CoordinatorParameters iparams, GroupElement Z, GroupElement Cx0, GroupElement Cx1)
+			=> new Statement(new GroupElement[,]
+			{
+				// public                     Witness terms:
+				// point      z               z0              t
+				{ Z,          iparams.I,      O,              O },
+				{ Cx1,        Generators.Gx1, Generators.Gx0, Cx0 }, // Cx1 = z*Gx1 + t*U, z0 cancels t*z*Gx0 out of t*Cx0 leaving z*Gx1 + t*U
+			});
+
```

Comment:
```suggestion

		// The coordinator derives Z from the secret key and the randomized commitments. If the
		// MacShow proof is valid, this implies that the MAC is valid, because only a CV point
		// generated using a valid MAC will cancel out the terms depending on the secret key here.
```